src/agent/Core/ApplicationPool/Group/ProcessListManagement.cpp

```diff
@@ -58,71 +58,81 @@ Group::findProcessWithStickySessionId(unsigned int id) const {
 }
 
 Process *
-Group::findProcessWithStickySessionIdOrLowestBusyness(unsigned int id) const {
-	int leastBusyProcessIndex = -1;
-	int lowestBusyness = 0;
-	unsigned int i, size = enabledProcessBusynessLevels.size();
-	const int *enabledProcessBusynessLevels = &this->enabledProcessBusynessLevels[0];
-
-	for (i = 0; i < size; i++) {
-		Process *process = enabledProcesses[i].get();
+Group::findBestProcessPreferringStickySessionId(unsigned int id) const {
+	Process *bestProcess = nullptr;
+	ProcessList::const_iterator it;
+	ProcessList::const_iterator end = enabledProcesses.end();
+	for (it = enabledProcesses.begin(); it != end; it++) {
+		Process *process = (*it).get();
 		if (process->getStickySessionId() == id) {
 			return process;
-		} else if (leastBusyProcessIndex == -1 || enabledProcessBusynessLevels[i] < lowestBusyness) {
-			leastBusyProcessIndex = i;
-			lowestBusyness = enabledProcessBusynessLevels[i];
+        } else if (bestProcess == nullptr ||
+                   process->generation > bestProcess->generation ||
+				   (process->generation == bestProcess->generation && process->spawnerCreationTime < bestProcess->spawnerCreationTime) ||
+				   (process->generation == bestProcess->generation && process->spawnerCreationTime == bestProcess->spawnerCreationTime && process->busyness() < bestProcess->busyness())
+        ) {
+			bestProcess = process;
 		}
 	}
-
-	if (leastBusyProcessIndex == -1) {
-		return NULL;
-	} else {
-		return enabledProcesses[leastBusyProcessIndex].get();
-	}
+	return bestProcess;
 }
 
 Process *
-Group::findProcessWithLowestBusyness(const ProcessList &processes) const {
+Group::findBestProcess(const ProcessList &processes) const {
 	if (processes.empty()) {
-		return NULL;
+		return nullptr;
 	}
 
-	int lowestBusyness = -1;
-	Process *leastBusyProcess = NULL;
+	Process *bestProcess = nullptr;
 	ProcessList::const_iterator it;
 	ProcessList::const_iterator end = processes.end();
 	for (it = processes.begin(); it != end; it++) {
 		Process *process = (*it).get();
-		int busyness = process->busyness();
-		if (lowestBusyness == -1 || lowestBusyness > busyness) {
-			lowestBusyness = busyness;
-			leastBusyProcess = process;
+
+		if (bestProcess == nullptr ||
+			process->generation > bestProcess->generation ||
+			(process->generation == bestProcess->generation && process->spawnerCreationTime < bestProcess->spawnerCreationTime) ||
+			(process->generation == bestProcess->generation && process->spawnerCreationTime == bestProcess->spawnerCreationTime && process->busyness() < bestProcess->busyness())
+        ) {
+			bestProcess = process;
 		}
 	}
-	return leastBusyProcess;
+	return bestProcess;
 }
 
 /**
- * Cache-optimized version of findProcessWithLowestBusyness() for the common case.
+ * Cache-optimized version of findBestProcess() for the common case.
  */
 Process *
-Group::findEnabledProcessWithLowestBusyness() const {
+Group::findBestEnabledProcess() const {
```

Comment:
My head is clearer today and I thought of a problem with the routing algorithm. Recall that the algorithm has now turned into essentially this:

> Select the first process according to the order `generation DESC, startTime ASC, busyness ASC`.

This means that during a restart, new requests are only routed to next-generation processes, while previous-generation processes are only allowed to finish their existing requests. That's pretty bad.

I suggest a two-phase routing algorithm:

1. First, we select the "best process" according to the algorithm as you have already coded in this branch.
2. If it turns out that this process `isTotallyBusy()`, then we fallback to the old algorithm where we select the least busy process amongst all generations. This way we won't waste the old processes.

What do you think?

Reply:
Makes more sense to me than the last algorithm.

Reply:
Also since we know we're going to punt if the process `isTotallyBusy()` there's no need to loop twice, just reject if `isTotallyBusy()` is true as we iterate.

It does raise another question though, what do we do if all processes are totally busy?

Reply:
I pushed a possible solution. I keep track of the best process which is totally busy, until I find one that is not totally busy, at the end of the loop I return either the best process which is not totally busy or if all processes are totally busy, then the best totally busy process.

The mechanism for choosing which process is best is decided as above for both processes, but tracked separately.

Reply:
> It does raise another question though, what do we do if all processes are totally busy?

Doesn't matter. The caller (`Group::route()`) will check whether returned process is totally busy. If yes then it'll take that as indication that the request needs to be queued until more capacity becomes available (either because of another request finishing, or by spawning more processes).

Reply:
> It does raise another question though, what do we do if all processes are totally busy?

Doesn't matter. The caller (`Group::route()`) will check whether returned process is totally busy. If yes then it'll take that as indication that the request needs to be queued until more capacity becomes available (either because of another request finishing, or by spawning more processes).

See also my other comment about the `if` block.